balancer/rls/internal/config.go

```diff
@@ -0,0 +1,149 @@
+	serviceconfig.LoadBalancingConfig
+
+	kbMap                keys.BuilderMap
+	lookupService        resolver.Target
+	lookupServiceTimeout time.Duration
+	maxAge               time.Duration
+	staleAge             time.Duration
+	cacheSizeBytes       int64
+	rpStrategy           rlspb.RouteLookupConfig_RequestProcessingStrategy
+	cpName               string
+	cpTargetField        string
+	cpConfig             map[string]json.RawMessage
+}
+
+// lbParsedConfig is an parsed internal representation of the JSON
+// loadBalancing config for the RLS LB policy. This config is put through
+// further validation checks and transformed into another representation before
+// handing it off to the RLS LB policy.
+type lbParsedConfig struct {
+	rlsProto    *rlspb.RouteLookupConfig
+	childPolicy *loadBalancingConfig
+	targetField string
+}
+
+// UnmarshalJSON parses the JSON-encoded byte slice in data and stores it in l.
+// When unmarshalling, we iterate through the childPolicy list and select the
+// first LB policy which has been registered.
+// Returns error only when JSON unmarshaling fails. Further validation checks
+// are required to verify the sanity of the parsed config.
+func (l *lbParsedConfig) UnmarshalJSON(data []byte) error {
+	var jsonData map[string]json.RawMessage
+	if err := json.Unmarshal(data, &jsonData); err != nil {
+		return fmt.Errorf("bad input json data: %v", err)
+	}
+
+	// We declare local variables to store the result of JSON unmarshaling, and
+	// set them to appropriate fields in the lbParsedConfig only if everything goes
+	// smoothly.
+	var (
+		rlsProto    *rlspb.RouteLookupConfig
+		childPolicy *loadBalancingConfig
+		targetField string
+	)
+	for k, v := range jsonData {
+		switch k {
+		case "routeLookupConfig":
+			m := jsonpb.Unmarshaler{AllowUnknownFields: true}
+			rlsProto = &rlspb.RouteLookupConfig{}
+			if err := m.Unmarshal(bytes.NewReader(v), rlsProto); err != nil {
+				return fmt.Errorf("bad routeLookupConfig proto: %v", err)
+			}
+		case "childPolicy":
+			var lbCfgs []*loadBalancingConfig
+			if err := json.Unmarshal(v, &lbCfgs); err != nil {
+				return fmt.Errorf("bad childPolicy config: %v", err)
+			}
+			for _, lbcfg := range lbCfgs {
+				if balancer.Get(lbcfg.Name) != nil {
+					childPolicy = lbcfg
+					break
+				}
+			}
+		case "childPolicyConfigTargetFieldName":
+			if err := json.Unmarshal(v, &targetField); err != nil {
+				return fmt.Errorf("bad childPolicyConfigTargetFieldName: %v", err)
+			}
+		default:
+			grpclog.Warningf("rls: unknown field %q in ServiceConfig", k)
+		}
+	}
+
+	l.rlsProto = rlsProto
+	l.childPolicy = childPolicy
+	l.targetField = targetField
+	return nil
+}
+
+// MarshalJSON returns a JSON encoding of l.
+func (l *lbParsedConfig) MarshalJSON() ([]byte, error) {
+	return nil, fmt.Errorf("lbParsedConfig.MarshalJSON() is unimplemented")
+}
+
+// loadBalancingConfig represents a single load balancing config,
+// stored in JSON format.
+type loadBalancingConfig struct {
```

Comment:
We do similar things for load balancing config lists in https://github.com/grpc/grpc-go/blob/master/service_config.go#L252.  Maybe we should factor this out and re-use it?